syncer/dml_worker.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	brutils "github.com/pingcap/tidb/br/pkg/utils"
+	"go.uber.org/zap"
+
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/terror"
+	"github.com/pingcap/dm/pkg/utils"
+	"github.com/pingcap/dm/syncer/dbconn"
+	"github.com/pingcap/dm/syncer/metrics"
+)
+
+// DMLWorker is used to sync dml.
+type DMLWorker struct {
+	batch          int
+	workerCount    int
+	chanSize       int
+	toDBConns      []*dbconn.DBConn
+	tctx           *tcontext.Context
+	causalityWg    sync.WaitGroup
+	connectionPool *brutils.WorkerPool
```

Comment:
maybe call this workerPool ?

Reply:
I think connectionPool it's better because now we can have 100 causality workers but only 10 db connections. This pool is actually a connection pool.